Maybe unsound in SkBuff

hello, thank you for your contribution in this project, I am scanning the unsoundness problem in rust project.
I notice the following code:
```
pub struct SkBuff {
    pub skb: *mut __sk_buff,
}


impl SkBuff {
    pub fn new(skb: *mut __sk_buff) -> SkBuff {
        SkBuff { skb }
    }

    #[allow(clippy::len_without_is_empty)]
    #[inline]
    pub fn len(&self) -> u32 {
        unsafe { (*self.skb).len }
    }
............................
}
```
Considering that `programs` is a pub mod and `skb` is a pub field, I assume that users can directly manipulate this field, and that `len` is a public function. This potential situation could result in `self.skb` being a null pointer, and directly dereferencing it might trigger undefined behavior (UB). For safety reasons, I felt it necessary to report this issue. If you have performed checks elsewhere that ensure this is safe, please don’t take offense at my raising this issue.
Sorry, I can't provide a Poc, because trying to import SkBuff in my environment will get an error:
```
error[E0432]: unresolved import `aya::programs::SkBuff`
 --> src/main.rs:3:5
  |
3 | use aya::programs::SkBuff;
  |     ^^^^^^^^^^^^^^^^^^^^^ no `SkBuff` in `programs`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `lwz` (bin "lwz") due to 1 previous error
```
If this can be solved, I am happy to provide a Poc.

ofcourse, here is my main.rs:
```
extern crate aya;

use aya::programs::SkBuff;
fn main() {
    
}
```
and Cargo.toml:
```
[package]
name = "lwz"
version = "0.1.0"
authors = ["lwz"]
edition = "2018"

[dependencies]
aya = "0.13.1"
```
result:
```
error[E0432]: unresolved import `aya::programs::SkBuff`
 --> src/main.rs:3:5
  |
3 | use aya::programs::SkBuff;
  |     ^^^^^^^^^^^^^^^^^^^^^ no `SkBuff` in `programs`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `lwz` (bin "lwz") due to 1 previous error
lwz@LAPTOP-F68E79VH:/mnt/e/Github/lwz$
```

--- a/aya-ebpf-macros/src/flow_dissector.rs
+++ b/aya-ebpf-macros/src/flow_dissector.rs
@@ -29,6 +29,8 @@ impl FlowDissector {
             #[unsafe(no_mangle)]
             #[unsafe(link_section = "flow_dissector")]
             #vis fn #fn_name(ctx: *mut ::aya_ebpf::bindings::__sk_buff) -> u32 {
+                // SAFETY: The `ctx` pointer provided by the kernel should be valid.
+                let ctx = unsafe { ::core::ptr::NonNull::new_unchecked(ctx) };
                 return #fn_name(::aya_ebpf::programs::FlowDissectorContext::new(ctx));
 
                 #item
@@ -59,6 +61,8 @@ mod tests {
             #[unsafe(no_mangle)]
             #[unsafe(link_section = "flow_dissector")]
             fn prog(ctx: *mut ::aya_ebpf::bindings::__sk_buff) -> u32 {
+                // SAFETY: The `ctx` pointer provided by the kernel should be valid.
+                let ctx = unsafe { ::core::ptr::NonNull::new_unchecked(ctx) };
                 return prog(::aya_ebpf::programs::FlowDissectorContext::new(ctx));
 
                 fn prog(ctx: &mut ::aya_ebpf::programs::FlowDissectorContext) -> u32 {

--- a/aya-ebpf-macros/src/sk_skb.rs
+++ b/aya-ebpf-macros/src/sk_skb.rs
@@ -52,6 +52,8 @@ impl SkSkb {
             #[unsafe(no_mangle)]
             #[unsafe(link_section = #section_name)]
             #vis fn #fn_name(ctx: *mut ::aya_ebpf::bindings::__sk_buff) -> u32 {
+                // SAFETY: The `ctx` pointer provided by the kernel should be valid.
+                let ctx = unsafe { ::core::ptr::NonNull::new_unchecked(ctx) };
                 return #fn_name(::aya_ebpf::programs::SkBuffContext::new(ctx));
 
                 #item
@@ -83,6 +85,8 @@ mod tests {
             #[unsafe(no_mangle)]
             #[unsafe(link_section = "sk_skb/stream_parser")]
             fn prog(ctx: *mut ::aya_ebpf::bindings::__sk_buff) -> u32 {
+                // SAFETY: The `ctx` pointer provided by the kernel should be valid.
+                let ctx = unsafe { ::core::ptr::NonNull::new_unchecked(ctx) };
                 return prog(::aya_ebpf::programs::SkBuffContext::new(ctx));
 
                 fn prog(ctx: &mut ::aya_ebpf::programs::SkBuffContext) -> u32 {
@@ -110,6 +114,8 @@ mod tests {
             #[unsafe(no_mangle)]
             #[unsafe(link_section = "sk_skb/stream_verdict")]
             fn prog(ctx: *mut ::aya_ebpf::bindings::__sk_buff) -> u32 {
+                // SAFETY: The `ctx` pointer provided by the kernel should be valid.
+                let ctx = unsafe { ::core::ptr::NonNull::new_unchecked(ctx) };
                 return prog(::aya_ebpf::programs::SkBuffContext::new(ctx));
 
                 fn prog(ctx: &mut ::aya_ebpf::programs::SkBuffContext) -> u32 {

--- a/aya-ebpf-macros/src/socket_filter.rs
+++ b/aya-ebpf-macros/src/socket_filter.rs
@@ -29,6 +29,8 @@ impl SocketFilter {
             #[unsafe(no_mangle)]
             #[unsafe(link_section = "socket")]
             #vis fn #fn_name(ctx: *mut ::aya_ebpf::bindings::__sk_buff) -> i64 {
+                // SAFETY: The `ctx` pointer provided by the kernel should be valid.
+                let ctx = unsafe { ::core::ptr::NonNull::new_unchecked(ctx) };
                 return #fn_name(::aya_ebpf::programs::SkBuffContext::new(ctx));
 
                 #item
@@ -59,6 +61,8 @@ mod tests {
             #[unsafe(no_mangle)]
             #[unsafe(link_section = "socket")]
             fn prog(ctx: *mut ::aya_ebpf::bindings::__sk_buff) -> i64 {
+                // SAFETY: The `ctx` pointer provided by the kernel should be valid.
+                let ctx = unsafe { ::core::ptr::NonNull::new_unchecked(ctx) };
                 return prog(::aya_ebpf::programs::SkBuffContext::new(ctx));
 
                 fn prog(ctx: &mut ::aya_ebpf::programs::SkBuffContext) -> i64 {

--- a/aya-ebpf-macros/src/tc.rs
+++ b/aya-ebpf-macros/src/tc.rs
@@ -29,6 +29,8 @@ impl SchedClassifier {
             #[unsafe(no_mangle)]
             #[unsafe(link_section = "classifier")]
             #vis fn #fn_name(ctx: *mut ::aya_ebpf::bindings::__sk_buff) -> i32 {
+                // SAFETY: The `ctx` pointer provided by the kernel should be valid.
+                let ctx = unsafe { ::core::ptr::NonNull::new_unchecked(ctx) };
                 return #fn_name(::aya_ebpf::programs::TcContext::new(ctx));
 
                 #item
@@ -59,6 +61,8 @@ mod tests {
             #[unsafe(no_mangle)]
             #[unsafe(link_section = "classifier")]
             fn prog(ctx: *mut ::aya_ebpf::bindings::__sk_buff) -> i32 {
+                // SAFETY: The `ctx` pointer provided by the kernel should be valid.
+                let ctx = unsafe { ::core::ptr::NonNull::new_unchecked(ctx) };
                 return prog(::aya_ebpf::programs::TcContext::new(ctx));
 
                 fn prog(ctx: &mut ::aya_ebpf::programs::TcContext) -> i32 {

--- a/ebpf/aya-ebpf/src/btf_maps/sock_hash.rs
+++ b/ebpf/aya-ebpf/src/btf_maps/sock_hash.rs
@@ -108,7 +108,7 @@ impl<K, const MAX_ENTRIES: usize, const FLAGS: usize> SockHash<K, MAX_ENTRIES, F
         let () = Self::_CHECK;
         unsafe {
             bpf_sk_redirect_hash(
-                ctx.borrow().skb.skb,
+                ctx.borrow().skb.as_raw_ptr(),
                 self.as_ptr().cast(),
                 ptr::from_mut(key.borrow_mut()).cast(),
                 flags,

--- a/ebpf/aya-ebpf/src/maps/sock_hash.rs
+++ b/ebpf/aya-ebpf/src/maps/sock_hash.rs
@@ -72,7 +72,7 @@ impl<K> SockHash<K> {
     ) -> c_long {
         unsafe {
             bpf_sk_redirect_hash(
-                ctx.borrow().skb.skb,
+                ctx.borrow().skb.as_raw_ptr(),
                 self.def.as_ptr().cast(),
                 ptr::from_mut(key.borrow_mut()).cast(),
                 flags,

--- a/ebpf/aya-ebpf/src/programs/flow_dissector.rs
+++ b/ebpf/aya-ebpf/src/programs/flow_dissector.rs
@@ -1,3 +1,5 @@
+use core::ptr::NonNull;
+
 use aya_ebpf_cty::{c_long, c_void};
 
 use crate::{
@@ -12,12 +14,8 @@ pub struct FlowDissectorContext {
 
 impl FlowDissectorContext {
     #[inline]
-    #[expect(
-        clippy::not_unsafe_ptr_arg_deref,
-        reason = "skb is initialization context from kernel"
-    )]
-    pub const fn new(skb: *mut __sk_buff) -> Self {
-        let skb = unsafe { SkBuff::new(skb) };
+    pub const fn new(skb: NonNull<__sk_buff>) -> Self {
+        let skb = SkBuff::new(skb);
         Self { skb }
     }
 

--- a/ebpf/aya-ebpf/src/programs/flow_dissector.rs
+++ b/ebpf/aya-ebpf/src/programs/flow_dissector.rs
@@ -11,24 +11,29 @@ pub struct FlowDissectorContext {
 }
 
 impl FlowDissectorContext {
+    #[inline]
+    #[expect(
+        clippy::not_unsafe_ptr_arg_deref,
+        reason = "skb is initialization context from kernel"
+    )]
     pub const fn new(skb: *mut __sk_buff) -> Self {
-        let skb = SkBuff { skb };
+        let skb = unsafe { SkBuff::new(skb) };
         Self { skb }
     }
 
     #[inline]
-    pub fn data(&self) -> usize {
+    pub const fn data(&self) -> usize {
         self.skb.data()
     }
 
     #[inline]
-    pub fn data_end(&self) -> usize {
+    pub const fn data_end(&self) -> usize {
         self.skb.data_end()
     }
 
     #[inline]
     pub fn flow_keys(&mut self) -> &mut bpf_flow_keys {
-        unsafe { &mut *(*self.skb.skb).__bindgen_anon_1.flow_keys }
+        unsafe { &mut *(*self.skb.as_raw_ptr()).__bindgen_anon_1.flow_keys }
     }
 
     #[inline(always)]

--- a/ebpf/aya-ebpf/src/programs/sk_buff.rs
+++ b/ebpf/aya-ebpf/src/programs/sk_buff.rs
@@ -18,17 +18,9 @@ pub struct SkBuff {
 }
 
 impl SkBuff {
-    /// Creates a new `SkBuff` from a raw `__sk_buff` pointer.
-    ///
-    /// # Safety
-    ///
-    /// The caller must ensure that `skb` points to a valid and initialized
-    /// `__sk_buff` structure managed by the kernel.
     #[inline]
-    pub const unsafe fn new(skb: *mut __sk_buff) -> Self {
-        Self {
-            skb: unsafe { NonNull::new_unchecked(skb) },
-        }
+    pub const fn new(skb: NonNull<__sk_buff>) -> Self {
+        Self { skb }
     }
 
     #[inline]
@@ -244,13 +236,9 @@ pub struct SkBuffContext {
 }
 
 impl SkBuffContext {
-    #[expect(
-        clippy::not_unsafe_ptr_arg_deref,
-        reason = "skb is initialization context from kernel"
-    )]
     #[inline]
-    pub const fn new(skb: *mut __sk_buff) -> Self {
-        let skb = unsafe { SkBuff::new(skb) };
+    pub const fn new(skb: NonNull<__sk_buff>) -> Self {
+        let skb = SkBuff::new(skb);
         Self { skb }
     }
 
@@ -345,17 +333,7 @@ impl SkBuffContext {
     /// ```
     #[inline(always)]
     pub fn load_bytes(&self, offset: usize, dst: &mut [u8]) -> Result<usize, c_long> {
-        let len = dst.len();
-        let len_u32 = u32::try_from(len).map_err(|core::num::TryFromIntError { .. }| -1)?;
-        let ret = unsafe {
-            bpf_skb_load_bytes(
-                self.skb.as_raw_ptr().cast(),
-                offset as u32,
-                dst.as_mut_ptr().cast(),
-                len_u32,
-            )
-        };
-        if ret == 0 { Ok(len) } else { Err(ret) }
+        self.skb.load_bytes(offset, dst)
     }
 
     #[inline]

--- a/ebpf/aya-ebpf/src/programs/sk_buff.rs
+++ b/ebpf/aya-ebpf/src/programs/sk_buff.rs
@@ -1,4 +1,8 @@
-use core::{ffi::c_void, mem::MaybeUninit, ptr};
+use core::{
+    ffi::c_void,
+    mem::MaybeUninit,
+    ptr::{self, NonNull},
+};
 
 use aya_ebpf_bindings::helpers::{
     bpf_clone_redirect, bpf_get_socket_uid, bpf_l3_csum_replace, bpf_l4_csum_replace,
@@ -10,49 +14,63 @@ use aya_ebpf_cty::c_long;
 use crate::{EbpfContext, bindings::__sk_buff};
 
 pub struct SkBuff {
-    pub skb: *mut __sk_buff,
+    skb: NonNull<__sk_buff>,
 }
 
 impl SkBuff {
-    pub const fn new(skb: *mut __sk_buff) -> Self {
-        Self { skb }
+    /// Creates a new `SkBuff` from a raw `__sk_buff` pointer.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `skb` points to a valid and initialized
+    /// `__sk_buff` structure managed by the kernel.
+    #[inline]
+    pub const unsafe fn new(skb: *mut __sk_buff) -> Self {
+        Self {
+            skb: unsafe { NonNull::new_unchecked(skb) },
+        }
     }
 
     #[inline]
-    pub fn len(&self) -> u32 {
-        unsafe { (*self.skb).len }
+    pub(crate) const fn as_raw_ptr(&self) -> *mut __sk_buff {
+        self.skb.as_ptr()
     }
 
     #[inline]
-    pub(crate) fn data(&self) -> usize {
-        unsafe { (*self.skb).data as usize }
+    pub(crate) const fn data(&self) -> usize {
+        unsafe { self.skb.as_ref().data as usize }
     }
 
     #[inline]
-    pub(crate) fn data_end(&self) -> usize {
-        unsafe { (*self.skb).data_end as usize }
+    pub(crate) const fn data_end(&self) -> usize {
+        unsafe { self.skb.as_ref().data_end as usize }
+    }
+
+    #[inline]
+    pub const fn len(&self) -> u32 {
+        unsafe { self.skb.as_ref().len }
     }
 
     #[inline]
     pub fn set_mark(&self, mark: u32) {
-        unsafe { (*self.skb).mark = mark }
+        unsafe { (*self.as_raw_ptr()).mark = mark }
     }
 
     #[inline]
     pub fn cb(&self) -> &[u32] {
-        unsafe { &(*self.skb).cb }
+        unsafe { &(*self.as_raw_ptr()).cb }
     }
 
     /// Returns a mutable slice to the control buffer (cb).
     #[inline]
     pub fn cb_mut(&mut self) -> &mut [u32] {
-        unsafe { &mut (*self.skb).cb }
+        unsafe { &mut (*self.as_raw_ptr()).cb }
     }
 
     /// Returns the owner UID of the socket associated to the SKB context.
     #[inline]
     pub fn get_socket_uid(&self) -> u32 {
-        unsafe { bpf_get_socket_uid(self.skb) }
+        unsafe { bpf_get_socket_uid(self.as_raw_ptr()) }
     }
 
     #[inline]
@@ -60,7 +78,7 @@ impl SkBuff {
         unsafe {
             let mut data = MaybeUninit::<T>::uninit();
             let ret = bpf_skb_load_bytes(
-                self.skb.cast(),
+                self.as_raw_ptr().cast(),
                 offset as u32,
                 ptr::from_mut(&mut data).cast(),
                 size_of_val(&data) as u32,
@@ -90,7 +108,7 @@ impl SkBuff {
         let len_u32 = u32::try_from(len).map_err(|core::num::TryFromIntError { .. }| -1)?;
         let ret = unsafe {
             bpf_skb_load_bytes(
-                self.skb.cast(),
+                self.as_raw_ptr().cast(),
                 offset as u32,
                 dst.as_mut_ptr().cast(),
                 len_u32,
@@ -103,7 +121,7 @@ impl SkBuff {
     pub fn store<T>(&self, offset: usize, v: &T, flags: u64) -> Result<(), c_long> {
         unsafe {
             let ret = bpf_skb_store_bytes(
-                self.skb.cast(),
+                self.as_raw_ptr(),
                 offset as u32,
                 ptr::from_ref(v).cast(),
                 size_of_val(v) as u32,
@@ -122,7 +140,7 @@ impl SkBuff {
         size: u64,
     ) -> Result<(), c_long> {
         unsafe {
-            let ret = bpf_l3_csum_replace(self.skb.cast(), offset as u32, from, to, size);
+            let ret = bpf_l3_csum_replace(self.as_raw_ptr(), offset as u32, from, to, size);
             if ret == 0 { Ok(()) } else { Err(ret) }
         }
     }
@@ -136,32 +154,32 @@ impl SkBuff {
         flags: u64,
     ) -> Result<(), c_long> {
         unsafe {
-            let ret = bpf_l4_csum_replace(self.skb.cast(), offset as u32, from, to, flags);
+            let ret = bpf_l4_csum_replace(self.as_raw_ptr(), offset as u32, from, to, flags);
             if ret == 0 { Ok(()) } else { Err(ret) }
         }
     }
 
     #[inline]
     pub fn adjust_room(&self, len_diff: i32, mode: u32, flags: u64) -> Result<(), c_long> {
-        let ret = unsafe { bpf_skb_adjust_room(self.skb, len_diff, mode, flags) };
+        let ret = unsafe { bpf_skb_adjust_room(self.as_raw_ptr(), len_diff, mode, flags) };
         if ret == 0 { Ok(()) } else { Err(ret) }
     }
 
     #[inline]
     pub fn clone_redirect(&self, if_index: u32, flags: u64) -> Result<(), c_long> {
-        let ret = unsafe { bpf_clone_redirect(self.skb, if_index, flags) };
+        let ret = unsafe { bpf_clone_redirect(self.as_raw_ptr(), if_index, flags) };
         if ret == 0 { Ok(()) } else { Err(ret) }
     }
 
     #[inline]
     pub fn change_proto(&self, proto: u16, flags: u64) -> Result<(), c_long> {
-        let ret = unsafe { bpf_skb_change_proto(self.skb, proto, flags) };
+        let ret = unsafe { bpf_skb_change_proto(self.as_raw_ptr(), proto, flags) };
         if ret == 0 { Ok(()) } else { Err(ret) }
     }
 
     #[inline]
     pub fn change_type(&self, ty: u32) -> Result<(), c_long> {
-        let ret = unsafe { bpf_skb_change_type(self.skb, ty) };
+        let ret = unsafe { bpf_skb_change_type(self.as_raw_ptr(), ty) };
         if ret == 0 { Ok(()) } else { Err(ret) }
     }
 
@@ -172,52 +190,52 @@ impl SkBuff {
     /// for reading and writing with direct packet access.
     #[inline(always)]
     pub fn pull_data(&self, len: u32) -> Result<(), c_long> {
-        let ret = unsafe { bpf_skb_pull_data(self.skb, len) };
+        let ret = unsafe { bpf_skb_pull_data(self.as_raw_ptr(), len) };
         if ret == 0 { Ok(()) } else { Err(ret) }
     }
 
     pub(crate) const fn as_ptr(&self) -> *mut c_void {
-        self.skb.cast()
+        self.skb.as_ptr().cast()
     }
 
     #[inline]
-    pub fn protocol(&self) -> u32 {
-        unsafe { (*self.skb).protocol }
+    pub const fn protocol(&self) -> u32 {
+        unsafe { self.skb.as_ref().protocol }
     }
 
     #[inline]
-    pub fn family(&self) -> u32 {
-        unsafe { (*self.skb).family }
+    pub const fn family(&self) -> u32 {
+        unsafe { self.skb.as_ref().family }
     }
 
     #[inline]
-    pub fn local_ipv4(&self) -> u32 {
-        unsafe { (*self.skb).local_ip4 }
+    pub const fn local_ipv4(&self) -> u32 {
+        unsafe { self.skb.as_ref().local_ip4 }
     }
 
     #[inline]
-    pub fn local_ipv6(&self) -> &[u32; 4] {
-        unsafe { &(*self.skb).local_ip6 }
+    pub const fn local_ipv6(&self) -> &[u32; 4] {
+        unsafe { &self.skb.as_ref().local_ip6 }
     }
 
     #[inline]
-    pub fn remote_ipv4(&self) -> u32 {
-        unsafe { (*self.skb).remote_ip4 }
+    pub const fn remote_ipv4(&self) -> u32 {
+        unsafe { self.skb.as_ref().remote_ip4 }
     }
 
     #[inline]
-    pub fn remote_ipv6(&self) -> &[u32; 4] {
-        unsafe { &(*self.skb).remote_ip6 }
+    pub const fn remote_ipv6(&self) -> &[u32; 4] {
+        unsafe { &self.skb.as_ref().remote_ip6 }
     }
 
     #[inline]
-    pub fn local_port(&self) -> u32 {
-        unsafe { (*self.skb).local_port }
+    pub const fn local_port(&self) -> u32 {
+        unsafe { self.skb.as_ref().local_port }
     }
 
     #[inline]
-    pub fn remote_port(&self) -> u32 {
-        unsafe { (*self.skb).remote_port }
+    pub const fn remote_port(&self) -> u32 {
+        unsafe { self.skb.as_ref().remote_port }
     }
 }
 
@@ -226,29 +244,46 @@ pub struct SkBuffContext {
 }
 
 impl SkBuffContext {
+    #[expect(
+        clippy::not_unsafe_ptr_arg_deref,
+        reason = "skb is initialization context from kernel"
+    )]
+    #[inline]
     pub const fn new(skb: *mut __sk_buff) -> Self {
-        let skb = SkBuff { skb };
+        let skb = unsafe { SkBuff::new(skb) };
         Self { skb }
     }
 
     #[inline]
-    pub fn len(&self) -> u32 {
+    pub const fn data(&self) -> usize {
+        self.skb.data()
+    }
+
+    #[inline]
+    pub const fn data_end(&self) -> usize {
+        self.skb.data_end()
+    }
+
+    #[inline]
+    pub const fn len(&self) -> u32 {
         self.skb.len()
     }
 
     #[inline]
     pub fn set_mark(&self, mark: u32) {
-        self.skb.set_mark(mark);
+        unsafe {
+            (*self.skb.as_raw_ptr()).mark = mark;
+        }
     }
 
     #[inline]
-    pub fn cb(&self) -> &[u32] {
-        self.skb.cb()
+    pub fn cb(&self) -> &[u32; 5] {
+        unsafe { &(*self.skb.as_raw_ptr()).cb }
     }
 
     #[inline]
-    pub fn cb_mut(&mut self) -> &mut [u32] {
-        self.skb.cb_mut()
+    pub fn cb_mut(&mut self) -> &mut [u32; 5] {
+        unsafe { &mut (*self.skb.as_raw_ptr()).cb }
     }
 
     /// Returns the owner UID of the socket associated to the SKB context.
@@ -310,7 +345,17 @@ impl SkBuffContext {
     /// ```
     #[inline(always)]
     pub fn load_bytes(&self, offset: usize, dst: &mut [u8]) -> Result<usize, c_long> {
-        self.skb.load_bytes(offset, dst)
+        let len = dst.len();
+        let len_u32 = u32::try_from(len).map_err(|core::num::TryFromIntError { .. }| -1)?;
+        let ret = unsafe {
+            bpf_skb_load_bytes(
+                self.skb.as_raw_ptr().cast(),
+                offset as u32,
+                dst.as_mut_ptr().cast(),
+                len_u32,
+            )
+        };
+        if ret == 0 { Ok(len) } else { Err(ret) }
     }
 
     #[inline]

--- a/ebpf/aya-ebpf/src/programs/tc.rs
+++ b/ebpf/aya-ebpf/src/programs/tc.rs
@@ -24,23 +24,28 @@ pub struct TcContext {
 }
 
 impl TcContext {
+    #[expect(
+        clippy::not_unsafe_ptr_arg_deref,
+        reason = "skb is initialization context from kernel"
+    )]
+    #[inline]
     pub const fn new(skb: *mut __sk_buff) -> Self {
-        let skb = SkBuff { skb };
+        let skb = unsafe { SkBuff::new(skb) };
         Self { skb }
     }
 
     #[inline]
-    pub fn len(&self) -> u32 {
+    pub const fn len(&self) -> u32 {
         self.skb.len()
     }
 
     #[inline]
-    pub fn data(&self) -> usize {
+    pub const fn data(&self) -> usize {
         self.skb.data()
     }
 
     #[inline]
-    pub fn data_end(&self) -> usize {
+    pub const fn data_end(&self) -> usize {
         self.skb.data_end()
     }
 
@@ -214,7 +219,8 @@ impl TcContext {
     #[inline]
     pub fn skb_under_cgroup<M: CgroupArrayMap>(&self, map: &M, index: u32) -> Result<bool, c_long> {
         // SAFETY: `self.skb.skb` and `map` are valid pointers managed by aya.
-        let ret = unsafe { bpf_skb_under_cgroup(self.skb.skb, map.as_ptr().cast(), index) };
+        let ret =
+            unsafe { bpf_skb_under_cgroup(self.skb.as_raw_ptr(), map.as_ptr().cast(), index) };
         match ret {
             1 => Ok(true),
             0 => Ok(false),

--- a/ebpf/aya-ebpf/src/programs/tc.rs
+++ b/ebpf/aya-ebpf/src/programs/tc.rs
@@ -1,3 +1,5 @@
+use core::ptr::NonNull;
+
 use aya_ebpf_cty::{c_long, c_void};
 
 use crate::{
@@ -24,13 +26,9 @@ pub struct TcContext {
 }
 
 impl TcContext {
-    #[expect(
-        clippy::not_unsafe_ptr_arg_deref,
-        reason = "skb is initialization context from kernel"
-    )]
     #[inline]
-    pub const fn new(skb: *mut __sk_buff) -> Self {
-        let skb = unsafe { SkBuff::new(skb) };
+    pub const fn new(skb: NonNull<__sk_buff>) -> Self {
+        let skb = SkBuff::new(skb);
         Self { skb }
     }
 

--- a/xtask/public-api/aya-ebpf.txt
+++ b/xtask/public-api/aya-ebpf.txt
@@ -1843,8 +1843,8 @@ impl core::panic::unwind_safe::UnwindSafe for aya_ebpf::programs::fexit::FExitCo
 pub mod aya_ebpf::programs::flow_dissector
 pub struct aya_ebpf::programs::flow_dissector::FlowDissectorContext
 impl aya_ebpf::programs::flow_dissector::FlowDissectorContext
-pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data(&self) -> usize
-pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data_end(&self) -> usize
+pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data(&self) -> usize
+pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data_end(&self) -> usize
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::flow_keys(&mut self) -> &mut aya_ebpf_bindings::x86_64::bindings::bpf_flow_keys
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
 pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
@@ -1961,7 +1961,6 @@ impl core::panic::unwind_safe::RefUnwindSafe for aya_ebpf::programs::retprobe::R
 impl core::panic::unwind_safe::UnwindSafe for aya_ebpf::programs::retprobe::RetProbeContext
 pub mod aya_ebpf::programs::sk_buff
 pub struct aya_ebpf::programs::sk_buff::SkBuff
-pub aya_ebpf::programs::sk_buff::SkBuff::skb: *mut aya_ebpf_bindings::x86_64::bindings::__sk_buff
 impl aya_ebpf::programs::sk_buff::SkBuff
 pub fn aya_ebpf::programs::sk_buff::SkBuff::adjust_room(&self, i32, u32, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuff::cb(&self) -> &[u32]
@@ -1969,22 +1968,22 @@ pub fn aya_ebpf::programs::sk_buff::SkBuff::cb_mut(&mut self) -> &mut [u32]
 pub fn aya_ebpf::programs::sk_buff::SkBuff::change_proto(&self, u16, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuff::change_type(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuff::clone_redirect(&self, u32, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::sk_buff::SkBuff::family(&self) -> u32
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::family(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuff::get_socket_uid(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuff::l3_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuff::l4_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::sk_buff::SkBuff::len(&self) -> u32
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::len(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuff::load<T>(&self, usize) -> core::result::Result<T, aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuff::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::sk_buff::SkBuff::local_ipv4(&self) -> u32
-pub fn aya_ebpf::programs::sk_buff::SkBuff::local_ipv6(&self) -> &[u32; 4]
-pub fn aya_ebpf::programs::sk_buff::SkBuff::local_port(&self) -> u32
-pub const fn aya_ebpf::programs::sk_buff::SkBuff::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
-pub fn aya_ebpf::programs::sk_buff::SkBuff::protocol(&self) -> u32
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::local_ipv4(&self) -> u32
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::local_ipv6(&self) -> &[u32; 4]
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::local_port(&self) -> u32
+pub unsafe const fn aya_ebpf::programs::sk_buff::SkBuff::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::protocol(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuff::pull_data(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::sk_buff::SkBuff::remote_ipv4(&self) -> u32
-pub fn aya_ebpf::programs::sk_buff::SkBuff::remote_ipv6(&self) -> &[u32; 4]
-pub fn aya_ebpf::programs::sk_buff::SkBuff::remote_port(&self) -> u32
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::remote_ipv4(&self) -> u32
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::remote_ipv6(&self) -> &[u32; 4]
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::remote_port(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuff::set_mark(&self, u32)
 pub fn aya_ebpf::programs::sk_buff::SkBuff::store<T>(&self, usize, &T, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 impl core::marker::Freeze for aya_ebpf::programs::sk_buff::SkBuff
@@ -1998,14 +1997,16 @@ pub struct aya_ebpf::programs::sk_buff::SkBuffContext
 pub aya_ebpf::programs::sk_buff::SkBuffContext::skb: aya_ebpf::programs::sk_buff::SkBuff
 impl aya_ebpf::programs::sk_buff::SkBuffContext
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::adjust_room(&self, i32, u32, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::sk_buff::SkBuffContext::cb(&self) -> &[u32]
-pub fn aya_ebpf::programs::sk_buff::SkBuffContext::cb_mut(&mut self) -> &mut [u32]
+pub fn aya_ebpf::programs::sk_buff::SkBuffContext::cb(&self) -> &[u32; 5]
+pub fn aya_ebpf::programs::sk_buff::SkBuffContext::cb_mut(&mut self) -> &mut [u32; 5]
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::change_type(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::clone_redirect(&self, u32, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
+pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::data(&self) -> usize
+pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::data_end(&self) -> usize
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::get_socket_uid(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::l3_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::l4_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::sk_buff::SkBuffContext::len(&self) -> u32
+pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::len(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::load<T>(&self, usize) -> core::result::Result<T, aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
 pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
@@ -2215,12 +2216,12 @@ pub fn aya_ebpf::programs::tc::TcContext::cb_mut(&mut self) -> &mut [u32]
 pub fn aya_ebpf::programs::tc::TcContext::change_proto(&self, u16, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::change_type(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::clone_redirect(&self, u32, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::tc::TcContext::data(&self) -> usize
-pub fn aya_ebpf::programs::tc::TcContext::data_end(&self) -> usize
+pub const fn aya_ebpf::programs::tc::TcContext::data(&self) -> usize
+pub const fn aya_ebpf::programs::tc::TcContext::data_end(&self) -> usize
 pub fn aya_ebpf::programs::tc::TcContext::get_socket_uid(&self) -> u32
 pub fn aya_ebpf::programs::tc::TcContext::l3_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::l4_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::tc::TcContext::len(&self) -> u32
+pub const fn aya_ebpf::programs::tc::TcContext::len(&self) -> u32
 pub fn aya_ebpf::programs::tc::TcContext::load<T>(&self, usize) -> core::result::Result<T, aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
 pub const fn aya_ebpf::programs::tc::TcContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
@@ -2389,8 +2390,8 @@ impl core::panic::unwind_safe::RefUnwindSafe for aya_ebpf::programs::fexit::FExi
 impl core::panic::unwind_safe::UnwindSafe for aya_ebpf::programs::fexit::FExitContext
 pub struct aya_ebpf::programs::FlowDissectorContext
 impl aya_ebpf::programs::flow_dissector::FlowDissectorContext
-pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data(&self) -> usize
-pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data_end(&self) -> usize
+pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data(&self) -> usize
+pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data_end(&self) -> usize
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::flow_keys(&mut self) -> &mut aya_ebpf_bindings::x86_64::bindings::bpf_flow_keys
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
 pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
@@ -2504,14 +2505,16 @@ pub struct aya_ebpf::programs::SkBuffContext
 pub aya_ebpf::programs::SkBuffContext::skb: aya_ebpf::programs::sk_buff::SkBuff
 impl aya_ebpf::programs::sk_buff::SkBuffContext
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::adjust_room(&self, i32, u32, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::sk_buff::SkBuffContext::cb(&self) -> &[u32]
-pub fn aya_ebpf::programs::sk_buff::SkBuffContext::cb_mut(&mut self) -> &mut [u32]
+pub fn aya_ebpf::programs::sk_buff::SkBuffContext::cb(&self) -> &[u32; 5]
+pub fn aya_ebpf::programs::sk_buff::SkBuffContext::cb_mut(&mut self) -> &mut [u32; 5]
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::change_type(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::clone_redirect(&self, u32, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
+pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::data(&self) -> usize
+pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::data_end(&self) -> usize
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::get_socket_uid(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::l3_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::l4_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::sk_buff::SkBuffContext::len(&self) -> u32
+pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::len(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::load<T>(&self, usize) -> core::result::Result<T, aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
 pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
@@ -2712,12 +2715,12 @@ pub fn aya_ebpf::programs::tc::TcContext::cb_mut(&mut self) -> &mut [u32]
 pub fn aya_ebpf::programs::tc::TcContext::change_proto(&self, u16, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::change_type(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::clone_redirect(&self, u32, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::tc::TcContext::data(&self) -> usize
-pub fn aya_ebpf::programs::tc::TcContext::data_end(&self) -> usize
+pub const fn aya_ebpf::programs::tc::TcContext::data(&self) -> usize
+pub const fn aya_ebpf::programs::tc::TcContext::data_end(&self) -> usize
 pub fn aya_ebpf::programs::tc::TcContext::get_socket_uid(&self) -> u32
 pub fn aya_ebpf::programs::tc::TcContext::l3_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::l4_csum_replace(&self, usize, u64, u64, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
-pub fn aya_ebpf::programs::tc::TcContext::len(&self) -> u32
+pub const fn aya_ebpf::programs::tc::TcContext::len(&self) -> u32
 pub fn aya_ebpf::programs::tc::TcContext::load<T>(&self, usize) -> core::result::Result<T, aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
 pub const fn aya_ebpf::programs::tc::TcContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self

--- a/xtask/public-api/aya-ebpf.txt
+++ b/xtask/public-api/aya-ebpf.txt
@@ -1847,7 +1847,7 @@ pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data(&sel
 pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data_end(&self) -> usize
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::flow_keys(&mut self) -> &mut aya_ebpf_bindings::x86_64::bindings::bpf_flow_keys
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
-pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
+pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::new(core::ptr::non_null::NonNull<aya_ebpf_bindings::x86_64::bindings::__sk_buff>) -> Self
 impl aya_ebpf::EbpfContext for aya_ebpf::programs::flow_dissector::FlowDissectorContext
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::as_ptr(&self) -> *mut aya_ebpf_cty::c_void
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::command(&self) -> core::result::Result<[u8; 16], i32>
@@ -1978,7 +1978,7 @@ pub fn aya_ebpf::programs::sk_buff::SkBuff::load_bytes(&self, usize, &mut [u8]) 
 pub const fn aya_ebpf::programs::sk_buff::SkBuff::local_ipv4(&self) -> u32
 pub const fn aya_ebpf::programs::sk_buff::SkBuff::local_ipv6(&self) -> &[u32; 4]
 pub const fn aya_ebpf::programs::sk_buff::SkBuff::local_port(&self) -> u32
-pub unsafe const fn aya_ebpf::programs::sk_buff::SkBuff::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
+pub const fn aya_ebpf::programs::sk_buff::SkBuff::new(core::ptr::non_null::NonNull<aya_ebpf_bindings::x86_64::bindings::__sk_buff>) -> Self
 pub const fn aya_ebpf::programs::sk_buff::SkBuff::protocol(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuff::pull_data(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub const fn aya_ebpf::programs::sk_buff::SkBuff::remote_ipv4(&self) -> u32
@@ -2009,7 +2009,7 @@ pub fn aya_ebpf::programs::sk_buff::SkBuffContext::l4_csum_replace(&self, usize,
 pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::len(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::load<T>(&self, usize) -> core::result::Result<T, aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
-pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
+pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::new(core::ptr::non_null::NonNull<aya_ebpf_bindings::x86_64::bindings::__sk_buff>) -> Self
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::pull_data(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::set_mark(&self, u32)
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::store<T>(&self, usize, &T, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
@@ -2224,7 +2224,7 @@ pub fn aya_ebpf::programs::tc::TcContext::l4_csum_replace(&self, usize, u64, u64
 pub const fn aya_ebpf::programs::tc::TcContext::len(&self) -> u32
 pub fn aya_ebpf::programs::tc::TcContext::load<T>(&self, usize) -> core::result::Result<T, aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
-pub const fn aya_ebpf::programs::tc::TcContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
+pub const fn aya_ebpf::programs::tc::TcContext::new(core::ptr::non_null::NonNull<aya_ebpf_bindings::x86_64::bindings::__sk_buff>) -> Self
 pub fn aya_ebpf::programs::tc::TcContext::pull_data(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::set_mark(&self, u32)
 pub fn aya_ebpf::programs::tc::TcContext::skb_under_cgroup<M: aya_ebpf::programs::tc::CgroupArrayMap>(&self, &M, u32) -> core::result::Result<bool, aya_ebpf_cty::od::c_long>
@@ -2394,7 +2394,7 @@ pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data(&sel
 pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::data_end(&self) -> usize
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::flow_keys(&mut self) -> &mut aya_ebpf_bindings::x86_64::bindings::bpf_flow_keys
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
-pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
+pub const fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::new(core::ptr::non_null::NonNull<aya_ebpf_bindings::x86_64::bindings::__sk_buff>) -> Self
 impl aya_ebpf::EbpfContext for aya_ebpf::programs::flow_dissector::FlowDissectorContext
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::as_ptr(&self) -> *mut aya_ebpf_cty::c_void
 pub fn aya_ebpf::programs::flow_dissector::FlowDissectorContext::command(&self) -> core::result::Result<[u8; 16], i32>
@@ -2517,7 +2517,7 @@ pub fn aya_ebpf::programs::sk_buff::SkBuffContext::l4_csum_replace(&self, usize,
 pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::len(&self) -> u32
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::load<T>(&self, usize) -> core::result::Result<T, aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
-pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
+pub const fn aya_ebpf::programs::sk_buff::SkBuffContext::new(core::ptr::non_null::NonNull<aya_ebpf_bindings::x86_64::bindings::__sk_buff>) -> Self
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::pull_data(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::set_mark(&self, u32)
 pub fn aya_ebpf::programs::sk_buff::SkBuffContext::store<T>(&self, usize, &T, u64) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
@@ -2723,7 +2723,7 @@ pub fn aya_ebpf::programs::tc::TcContext::l4_csum_replace(&self, usize, u64, u64
 pub const fn aya_ebpf::programs::tc::TcContext::len(&self) -> u32
 pub fn aya_ebpf::programs::tc::TcContext::load<T>(&self, usize) -> core::result::Result<T, aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::load_bytes(&self, usize, &mut [u8]) -> core::result::Result<usize, aya_ebpf_cty::od::c_long>
-pub const fn aya_ebpf::programs::tc::TcContext::new(*mut aya_ebpf_bindings::x86_64::bindings::__sk_buff) -> Self
+pub const fn aya_ebpf::programs::tc::TcContext::new(core::ptr::non_null::NonNull<aya_ebpf_bindings::x86_64::bindings::__sk_buff>) -> Self
 pub fn aya_ebpf::programs::tc::TcContext::pull_data(&self, u32) -> core::result::Result<(), aya_ebpf_cty::od::c_long>
 pub fn aya_ebpf::programs::tc::TcContext::set_mark(&self, u32)
 pub fn aya_ebpf::programs::tc::TcContext::skb_under_cgroup<M: aya_ebpf::programs::tc::CgroupArrayMap>(&self, &M, u32) -> core::result::Result<bool, aya_ebpf_cty::od::c_long>

